sdk/search/Azure.Search.Documents/src/Batching/Publisher.cs

```diff
@@ -422,12 +422,14 @@ private async Task PublishAsync(bool flush, CancellationToken cancellationToken)
             // Returns whether the batch is now full.
             bool FillBatchFromQueue(List<PublisherAction<T>> batch, Queue< PublisherAction<T>> queue)
             {
-                // TODO: Consider tracking the keys in the batch and requiring
-                // them to be unique to avoid error alignment problems
+                HashSet<string> documentIdentifiers = new HashSet<string>(StringComparer.Ordinal);
 
                 while (queue.Count > 0)
                 {
-                    if (batch.Count < BatchActionCount)
+                    // Stop filling the batch if we run into an action for a document that is already in the batch.
+                    // We want to keep the actions ordered and map any errors accurately to the documents,
+                    // so we need to split the batch on encountering an action for a previously queued document.
+                    if ((batch.Count < BatchActionCount) && documentIdentifiers.Add(queue.Peek().Key))
```

Comment:
While it should be O(1) operation, I'd move the `documentIdentifiers` line out and dequeue like the code did before, check it, then add it to the batch if not already in the queue. For a hot path, this should be a little faster.

Reply:
Not sure how that would work. Once we deque, that element cannot be pushed back in place. That's why the `Peek()` check is before we attempt to dequeue.

Reply:
```csharp
if (batch.Count < BatchActionCount)
{
  var action = queue.Dequeue();
  if (documentIdentifiers.Add(action.Key))
  {
    batch.Add(action);
  }
}

But perhaps I missed the intent in the comment about keeping things ordered. You're right, then: it needs to remain to keep results ordered with actions.

Reply:
Yes, ordering needs to be maintained. Even otherwise, the suggested change needs to have the else to put the action at the end of the queue. So, there's no benefit for the additional work.